File: config/src/utils/tableUtils.lua
-- Copyright 2018 Yahoo Inc.
-- Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.

--- a module provides util functions for table config.
-- @module tableUtils

local M = {}

local misc = require 'util/misc'

-------------------------------------------------------------------------------
-- Build Config
-------------------------------------------------------------------------------

--- Add physical table configs and logical table configs into table t.
--
-- @param tables A table containing two keys: 
--  physical - A table of physical table configuration, keyed on name
--  logical - A table of logical table configuration, keyed on name
-- @return The table for storing table configs and ready be parsed into json
function M.build_table_config(tables)

    local configuration = {
        physical = {},
        logical = {}
    }

    for name, physical_table in pairs(tables.physical) do
        local copy = misc.shallow_copy(physical_table)
        copy.name = copy.name or name
        copy.description = copy.description or name
        table.insert(configuration.physical, copy)
    end

    for name, logical_table in pairs(tables.logical) do
        local copy = misc.shallow_copy(logical_table)
        copy.name = copy.name or name
        copy.description = copy.description or name
        table.insert(configuration.logical, copy)
    end
    return t
end

return M

File: fili-core/src/main/java/com/yahoo/bard/webservice/table/TableUtils.java
// Copyright 2019 Oath Inc.
// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
package com.yahoo.bard.webservice.table;

import com.yahoo.bard.webservice.table.availability.Availability;
import com.yahoo.bard.webservice.util.SimplifiedIntervalList;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Utility methods to reduce code duplication in table implementations.
 */
public final class TableUtils {

    public static final Function<Stream<Map<?, SimplifiedIntervalList>>, Map<?, SimplifiedIntervalList>>
            ALL_INTERVALS_MERGER = mapStream -> mapStream.map(Map::entrySet)
                    .flatMap(Set::stream)
            .collect(Collectors.toMap(Map.Entry::getKey,
                    Map.Entry::getValue,
                    SimplifiedIntervalList::union,
                    HashMap::new));

    /**
     * Private constructor for utility class.
     */
    private TableUtils() {

    }

    /**
     * Merge all the intervals in a stream of Physical Tables returning a map from column to union of available
     * intervals.
     *
     * @param tableStream  The tables being unioned.
     *
     * @return  A map of columns to intervals.
     */
    @SuppressWarnings(value = "unchecked")
    public static Map<Column, SimplifiedIntervalList> unionMergeTableIntervals(
            Stream<? extends PhysicalTable> tableStream
    ) {
        return (Map<Column, SimplifiedIntervalList>) ALL_INTERVALS_MERGER.apply(
                tableStream.map(PhysicalTable::getAllAvailableIntervals)
        );
    }

    /**
     * Merge all the intervals in a stream of availabilities returning a map from physical column name to union of
     * available intervals.
     *
     * @param availabilityStream  The availabilities being unioned.
     *
     * @return  A map of physical column name to intervals.
     */
    @SuppressWarnings(value = "unchecked")
    public static Map<String, SimplifiedIntervalList> unionMergeAvailabilityIntervals(
            Stream<Availability> availabilityStream
    ) {
        return (Map<String, SimplifiedIntervalList>) ALL_INTERVALS_MERGER.apply(
                availabilityStream.map(Availability::getAllAvailableIntervals)
        );
    }
}
